File: Data Structures & Algorithms/islands-and-treasure/submission-0.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        if grid == []:
            return grid

        def update(i, j, curr=1):
            if i >= len(grid) or i<0 or j>=len(grid[0]) or j<0 or grid[i][j] == -1 or grid[i][j] == -1:
                return
            if grid[i][j] > curr:
                grid[i][j] = curr
            else:
                return
        
            curr += 1
            update(i+1, j, curr)
            update(i-1, j, curr)
            update(i, j+1, curr)
            update(i, j-1, curr)
            
        
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 0:
                    update(i+1, j)
                    update(i-1, j)
                    update(i, j+1)
                    update(i, j-1)

File: Data Structures & Algorithms/path-with-minimum-effort/submission-0.cpp
class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        int rows = heights.size();
        int cols = heights[0].size();
        
        // Min-heap stores: {current_max_effort, {x, y}}
        priority_queue<pair<int, pair<int, int>>, 
                       vector<pair<int, pair<int, int>>>, 
                       greater<pair<int, pair<int, int>>>> pq;
        
        // Distance table to track the minimum effort to reach each cell
        vector<vector<int>> dist(rows, vector<int>(cols, INT_MAX));
        
        pq.push({0, {0, 0}});
        dist[0][0] = 0;
        
        // Up, Down, Right, Left
        vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
        
        while (!pq.empty()) {
            auto [effort, pos] = pq.top();
            int x = pos.first;
            int y = pos.second;
            pq.pop();
            
            // If we reached the bottom-right, we are done
            if (x == rows - 1 && y == cols - 1) {
                return effort;
            }
            
            // Skip if we already found a better path to this specific cell
            if (effort > dist[x][y]) {
                continue;
            }
            
            // Explore neighbors
            for (auto& dir : directions) {
                int nx = x + dir.first;
                int ny = y + dir.second;
                
                // Check bounds
                if (nx >= 0 && nx < rows && ny >= 0 && ny < cols) {
                    // The effort to the neighbor is the max of the current path's effort 
                    // and the absolute difference in height to the neighbor
                    int next_effort = max(effort, abs(heights[nx][ny] - heights[x][y]));
                    
                    // If this new path is better, update and push to queue
                    if (next_effort < dist[nx][ny]) {
                        dist[nx][ny] = next_effort;
                        pq.push({next_effort, {nx, ny}});
                    }
                }
            }
        }
        return 0; // Fallback (shouldn't be reached if grid is valid)
    }
};